webapi/Plugins/Chat/ChatPlugin.cs

```diff
@@ -334,41 +251,41 @@ private async Task<CopilotChatMessage> GetChatResponseAsync(string chatId, strin
             await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting audience", cancellationToken);
             audience = await AsyncUtils.SafeInvokeAsync(
                 () => this.GetAudienceAsync(chatContext, cancellationToken), nameof(GetAudienceAsync));
-            chatHistory.AddSystemMessage(audience);
+            metaPrompt.AddSystemMessage(audience);
         }
 
         // Extract user intent from the conversation history.
         await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting user intent", cancellationToken);
         var userIntent = await AsyncUtils.SafeInvokeAsync(
             () => this.GetUserIntentAsync(chatContext, cancellationToken), nameof(GetUserIntentAsync));
-        chatHistory.AddSystemMessage(userIntent);
+        metaPrompt.AddSystemMessage(userIntent);
 
-        // Calculate the remaining token budget.
-        var remainingTokenBudget = this.GetChatContextTokenLimit(chatHistory, userMessage.ToFormattedString());
+        // Calculate max amount of tokens to use for memories
+        int maxRequestTokenBudget = this.GetMaxRequestTokenBudget();
+        int tokensUsed = TokenUtils.GetContextMessagesTokenCount(metaPrompt);
+        int chatMemoryTokenBudget = maxRequestTokenBudget
+            - tokensUsed
+            - TokenUtils.GetContextMessageTokenCount(AuthorRole.User, userMessage.ToFormattedString());
```

Comment:
I think this would make more sense in the `GetMaxRequestTokenBudget` function since it always has to be accounted for. Here, it feels like it was just thrown in randomly and only accounted for when considering chat memory token budget (not the overall token budget)

or rename this variable to something like `remainingTokenBudget` to not tie it to chat memory

Reply:
Actually that is precisely the problem: the user message should be considered ONLY for the memory budget but NOT for the overall budget.

The reason we DO want to to account for this in the memory budget is so that we ensure we later have room for the user message in our request.

The reason we DON'T want it in the overall budget is because it's the overall budget that's used to determine how many messages from the history we'll tack on to our request.

Remember that at this point, the user message has been saved to storage but **is not yet in the request** payload.

To illustrate how including the token count for the usage in the overall budget was a bug and that the fix is to keep it only for the memory budget, let's use an example.

Say you have a user message with a token count of 100.
Let's also say that the overall budget left after the system prompt, audience and user intent are tallied is 150.

Now, if you remove the user message token count (100) from the overall budget, you are left with only 50 tokens to use. Then, when we start adding chat history message to the request, we WON'T be able to add the most recent one (the user message with a size of 100) and we will send the request with no chat history messages whatsoever!

However, with this fix, we DON'T remove the user message token count (100) from the overall budget so we still have 150 tokens left. Then, when we add chat history messages to our request, we can add the most recent one (the user message with a size of 100) and up to an additional 50 tokens worth of messages.

Reply:
I see I see, thanks for explaining in such detail!

Reply:
Also great catch